internal/models/openai_embedding.go

```diff
@@ -0,0 +1,198 @@
+	Dimensions int `json:"dimensions,omitempty"`
+}
+
+type Usage struct {
+	// The number of tokens used by the prompt.
+	PromptTokens int `json:"prompt_tokens"`
+
+	// The total number of tokens used by the request.
+	TotalTokens int `json:"total_tokens"`
+}
+
+type EmbeddingData struct {
+	// The object type, which is always "embedding".
+	Object string `json:"object"`
+
+	// The embedding vector, which is a list of floats.
+	Embedding []float32 `json:"embedding"`
+
+	// The index of the embedding in the list of embeddings.
+	Index int `json:"index"`
+}
+
+type EmbeddingResponse struct {
+	// The object type, which is always "list".
+	Object string `json:"object"`
+
+	// The list of embeddings generated by the model.
+	Data []EmbeddingData `json:"data"`
+
+	// The name of the model used to generate the embedding.
+	Model string `json:"model"`
+
+	// The usage information for the request.
+	Usage Usage `json:"usage"`
+}
+
+type ByIndex struct {
+	resp *EmbeddingResponse
+}
+
+func (eb *ByIndex) Len() int { return len(eb.resp.Data) }
+func (eb *ByIndex) Swap(i, j int) {
+	eb.resp.Data[i], eb.resp.Data[j] = eb.resp.Data[j], eb.resp.Data[i]
+}
+func (eb *ByIndex) Less(i, j int) bool { return eb.resp.Data[i].Index < eb.resp.Data[j].Index }
+
+type ErrorInfo struct {
+	Code    string `json:"code"`
+	Message string `json:"message"`
+	Param   string `json:"param,omitempty"`
+	Type    string `json:"type"`
+}
+
+type EmbedddingError struct {
+	Error ErrorInfo `json:"error"`
+}
+
+type OpenAIEmbeddingClient struct {
+	apiKey string
+	url    string
+}
+
+func (c *OpenAIEmbeddingClient) Check() error {
+	if c.apiKey == "" {
+		return fmt.Errorf("OpenAI api key is empty")
+	}
+
+	if c.url == "" {
+		return fmt.Errorf("OpenAI embedding url is empty")
+	}
+	return nil
+}
+
+func NewOpenAIEmbeddingClient(apiKey string, url string) OpenAIEmbeddingClient {
+	return OpenAIEmbeddingClient{
+		apiKey: apiKey,
+		url:    url,
+	}
+}
+
+func (c *OpenAIEmbeddingClient) send(client *http.Client, req *http.Request, res *EmbeddingResponse) error {
+	// call openai
+	resp, err := client.Do(req)
+
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+
+	body, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return err
+	}
+
+	if resp.StatusCode != 200 {
+		return fmt.Errorf(string(body))
+	}
+
+	err = json.Unmarshal(body, &res)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (c *OpenAIEmbeddingClient) sendWithRetry(client *http.Client, req *http.Request, res *EmbeddingResponse, maxRetries int) error {
+	var err error
+	for i := 0; i < maxRetries; i++ {
+		err = c.send(client, req, res)
+		if err == nil {
+			return nil
+		}
+	}
+	return err
+}
+
+func (c *OpenAIEmbeddingClient) Embedding(modelName string, texts []string, dim int, user string, timeoutSec time.Duration) (*EmbeddingResponse, error) {
+	var r EmbeddingRequest
+	r.Model = modelName
+	r.Input = texts
+	r.EncodingFormat = "float"
+	if user != "" {
+		r.User = user
+	}
+	if dim != 0 {
```

Comment:
what if dim==0?

Reply:
The model itself will have default values